### PR TITLE
Laboratory work #3, Vladislava Tsvetkova - 22FPL2

### PR DESCRIPTION
**Pull request by Vladays**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -26,7 +24,7 @@ def __init__(self, end_of_word_token: str) -> None:
             end_of_word_token (str): A token denoting word boundary
         """
         self._end_of_word_token = end_of_word_token
-        self._storage = {end_of_word_token: 0}
+        self._storage = {'_': 0}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:27]: What `'_'` is ?

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -103,15 +102,10 @@ def get_token(self, element_id: int) -> Optional[str]:
...
+        if not isinstance(element, str) or element not in self._storage:
             return None
 
-        filtered_items = filter(lambda item: item[1] == element_id, self._storage.items())
-        token = next(filtered_items, None)
-        if token:
-            return token[0]
-
-        return None
+        return self._storage[element]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:108]: are you sure?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -182,33 +171,31 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
...
 
-        for token in (char for n_gram in content['freq']
-                      for char in n_gram.lower() if char.isalpha()):
-            self._put(token)
-
+        for element in content['freq']:
+            for token in element:
+                if token.isalpha():
+                    self._put(token)
+                if token in (' ', self._end_of_word_token):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:197]: you create it on each iteratiuon. Why?

File: lab_3_generate_by_ngrams/target_score.txt [hunk 4]
```diff
@@ -1 +1 @@
-10
+0
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/target_score.txt:1]: ?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/target_score.txt (hunk 4 above)
Comment (Vladays) [reply lab_3_generate_by_ngrams/target_score.txt:1]: changed

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -45,24 +43,26 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            text_list += word_list
+            text_list.append(self._end_of_word_token)
 
-        if text[-1].isdigit() or text[-1].isalpha():
-            tokens.pop(-1)
+        if text_list.count(self._end_of_word_token) == len(text_list):
+            return None
 
-        return tuple(tokens)
+        if text[-1].isalnum():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:63]: strange...

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -105,13 +104,9 @@ def get_token(self, element_id: int) -> Optional[str]:
         """
         if not isinstance(element_id, int):
             return None
+        if element_id not in self._storage.values()):
+            return None
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:108]: where is the imlementation?

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:198]: Create it once before loop

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -272,6 +260,7 @@ class NGramLanguageModel:
         _encoded_corpus (tuple): Encoded text
     """
 
+
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:263]: remove

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -430,31 +401,24 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(encoded)
+            if not next_tokens:
                 break
 
-            max_freq = max(tokens.values())
-            max_freq_tokens = [token for token, freq in tokens.items() if freq == max_freq]
-            max_freq_tokens = sorted(max_freq_tokens, reverse=True)
-            encoded_prompt += (max_freq_tokens[0],)
+            best_predictions = [token for token, freq in next_tokens.items() if freq == max(next_tokens.values())]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:418]: You can calculate `max` only once, not in the loop

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:108]: You removed it. Right now only checks  are on the function

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -319,28 +305,20 @@ def build(self) -> int:
...
-        for ngram, freq in context_freq_dict.items():
-            context = ngram[:-1]
-            lower_ngram_counts[context] = lower_ngram_counts.get(context, 0) + freq
-
-        self._n_gram_frequencies = {ngram: freq / lower_ngram_counts[ngram[:-1]]
-                                    for ngram, freq in context_freq_dict.items()}
-
+        for n_gram in set(n_grams):
+            number_of_n_grams = n_grams.count(n_gram)
+            context_count = len([context for context in n_grams
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:316]: use `Count`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (Vladays) [reply lab_3_generate_by_ngrams/main.py:108]: it's here now

**Pull request closed without merging**